receiver/httpdreceiver/scraper.go

```diff
@@ -0,0 +1,248 @@
+	}
+	serverName := u.Hostname()
+
+	for metricKey, metricValue := range parseStats(stats) {
+		labels := pdata.NewAttributeMap()
+		labels.Insert(metadata.L.ServerName, pdata.NewAttributeValueString(serverName))
+
+		switch metricKey {
+		case "ServerUptimeSeconds":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(uptime, labels, i, now)
+			}
+		case "ConnsTotal":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(connections, labels, i, now)
+			}
+		case "BusyWorkers":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				labels.Insert(metadata.L.WorkersState, pdata.NewAttributeValueString("busy"))
+				addToIntMetric(workers, labels, i, now)
+			}
+		case "IdleWorkers":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				labels.Insert(metadata.L.WorkersState, pdata.NewAttributeValueString("idle"))
+				addToIntMetric(workers, labels, i, now)
+			}
+		case "Total Accesses":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				addToIntMetric(requests, labels, i, now)
+			}
+		case "Total kBytes":
+			if i, ok := r.parseInt(metricKey, metricValue); ok {
+				bytes := kbytesToBytes(i)
+				addToIntMetric(traffic, labels, bytes, now)
+			}
+		case "Scoreboard":
+			scoreboardMap := parseScoreboard(metricValue)
+			for identifier, score := range scoreboardMap {
+				labels.Upsert(metadata.L.ScoreboardState, pdata.NewAttributeValueString(identifier))
+				addToIntMetric(scoreboard, labels, score, now)
+			}
+		}
+	}
+
+	return rms, nil
+}
+
+// GetStats collects metric stats by making a get request at an endpoint.
+func (r *httpdScraper) GetStats() (string, error) {
+	url := fmt.Sprintf("%s%s", r.cfg.Endpoint, "/server-status?auto")
+	resp, err := r.httpClient.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	return string(body), nil
+}
+
+// parseStats converts a response body key:values into a map.
+func parseStats(resp string) map[string]string {
+	metrics := make(map[string]string)
+
+	fields := strings.Split(resp, "\n")
+	for _, field := range fields {
+		index := strings.Index(field, ": ")
+		if index == -1 {
+			continue
+		}
+		metrics[field[:index]] = field[index+2:]
+	}
+	return metrics
+}
+
+// parseInt converts string to int64.
+func (r *httpdScraper) parseInt(key, value string) (int64, bool) {
+	i, err := strconv.ParseInt(value, 10, 64)
+	if err != nil {
+		r.logInvalid("int", key, value)
+		return 0, false
+	}
+	return i, true
+}
+
+func (r *httpdScraper) logInvalid(expectedType, key, value string) {
+	r.logger.Info(
+		"invalid value",
+		zap.String("expectedType", expectedType),
+		zap.String("key", key),
+		zap.String("value", value),
+	)
+}
+
+// parseScoreboard quantifies the symbolic mapping of the scoreboard.
+func parseScoreboard(values string) map[string]int64 {
+	scoreboard := map[string]int64{
```

Comment:
Consider using a (package-local) struct for this instead of a dictionary.

Reply:
Personally, I prefer the current implementation in this case. All the magic strings relating to scoreboard values are neatly encapsulated in this method. Defining a struct just to hold the counts would still leave the need for associating each field with a corresponding string, which calls for additional structure. For a simple string parsing routine like this, it seems like unnecessary abstraction.

That said, the meaning of the return value does require some interpretation. As such, I've created an alias to make this more self-describing.